### PR TITLE
New Dockerfile

### PR DESCRIPTION
**Pull request by squeaky-pl**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,6 +1,7 @@
-FROM ubuntu:16.04
+FROM ubuntu:xenial-20210804
```
Comment (squeaky-pl) [inline Dockerfile:1]: It felt appropriate to pin this, as this is new base image

File: .dockerignore [hunk 2]
```diff
@@ -0,0 +1,6 @@
+.circleci
```
Comment (squeaky-pl) [inline .dockerignore:1]: Added this so the layer caching works nicely

File: Dockerfile [hunk 3]
```diff
@@ -30,11 +32,31 @@ RUN DEBIAN_FRONTEND=noninteractive && apt-get update && apt-get dist-upgrade -y
...
+WORKDIR /opt/app
+
+ENV PATH="/opt/venv/bin:$PATH"
+
 COPY ./ ./
-RUN pip install setuptools==44.0.0
-RUN pip install pip==20.3.4
-RUN pip install -r requirements_frozen.txt -e .
+RUN \
+  virtualenv /opt/venv && \
```
Comment (AlecRosenbaum) [inline Dockerfile:56]: why use a virtualenv inside a docker image?

Comment (squeaky-pl) [reply Dockerfile:56]: This was a suggestion from @jkemp101

File: docker-compose.yml [hunk 4]
```diff
@@ -15,12 +15,13 @@ services:
...
 
-  nylas:
+  app:
     build:
       context: .
       dockerfile: Dockerfile
     volumes:
-      - .:/src
+      - .:/opt/app
+    tmpfs: /opt/app/inbox_sync.egg-info
```
Comment (squeaky-pl) [inline docker-compose.yml:24]: I had to do this so `pip install -e .` keeps working after mounting the source code volume now that we use non-root user and Docker on Linux mounts volumes as root: https://github.com/moby/moby/issues/2259

Another alternative would be to chown to 5000:5000 the code on the host but I found that ugly.

File: .circleci/config.yml [hunk 5]
```diff
@@ -54,6 +68,4 @@ workflows:
       - static-code-analysis:
           context:
             - docker-hub-creds-ro
-      - build:
-          context:
```
Comment (squeaky-pl) [inline .circleci/config.yml:58]: Had to remove this because it shadows the variables added on project level.

File: .circleci/config.yml [hunk 6]
```diff
@@ -28,24 +28,38 @@ jobs:
...
           name: run tests
           command: |
             docker login -u $DOCKER_USER -p $DOCKER_PASS
-            docker-compose run nylas bash -ec '
+            docker-compose run app bash -ec '
               pip install -e . \
               && bin/wait-for-it.sh mysql:3306 \
               && NYLAS_ENV=test py.test tests/
             '
+      - run:
```
Comment (squeaky-pl) [inline .circleci/config.yml:49]: Copied this from https://github.com/closeio/authalligator/blob/master/.circleci/config.yml#L98-L113

File: .circleci/config.yml (hunk 5 above)
Comment (AlecRosenbaum) [reply .circleci/config.yml:58]: should we remove it for `static-code-analysis` too?

File: .circleci/config.yml [hunk 7]
```diff
@@ -28,24 +28,38 @@ jobs:
...
+          name: Push the tested image to docker hub
+          environment:
+            PROJECT: sync-engine
+            PROJECT_NAMESPACE: closeio
+          command: |
+            PROJECT_VSN="$CIRCLE_BRANCH"
+            PROJECT_SHA="$CIRCLE_SHA1"
+            docker login -u $DOCKER_USER -p $DOCKER_PASS
+            # tag and push both sync-engine:master and sync-engine:<sha1>
+            docker tag "${PROJECT}_app" "${PROJECT_NAMESPACE}/${PROJECT}:branch-${PROJECT_VSN}"
```
Comment (AlecRosenbaum) [inline .circleci/config.yml:59]: where are `PROJECT` and `PROJECT_NAMESPACE` set?

Comment (squeaky-pl) [reply .circleci/config.yml:59]: @AlecRosenbaum not sure if I understand correctly but it's set under `enviornment:` couple of lines above.

Comment (AlecRosenbaum) [reply .circleci/config.yml:59]: 🤦‍♂️ Yep, my bad.

**Review by AlecRosenbaum — APPROVED** [review]

File: Dockerfile [hunk 8]
```diff
@@ -1,6 +1,7 @@
-FROM ubuntu:16.04
+FROM ubuntu:xenial-20210804
 
-USER root
+RUN groupadd -g 5000 sync-engine \
+  && useradd -d /home/sync-engine -m -u 5000 -g 5000 sync-engine
```
[the remark this answers is not in the document]
Comment (squeaky-pl) [reply Dockerfile:4]: @jkemp101 Is this a concern really? I'd rather not introduce another symbol. A typical Debian install has NAME_REGEX="^[a-z][-a-z0-9_]*\$" and dashes were allowed for example `www-data` is commonly used for Nginx and Apache.

[the remark this answers is not in the document]
Comment (squeaky-pl) [reply Dockerfile:4]: This specific Dockerfile won't run on any *NIX anyway because it contains references to Debian specific package names.

**Review by jkemp101 — APPROVED** [review]

**Pull request merged**